typescript/packages/jumble/src/components/CharmRunner.tsx

```diff
@@ -79,10 +80,19 @@ export function CharmRenderer({ charm, className = "" }: CharmRendererProps) {
     const container = containerRef.current;
     if (!container) return;
 
-    const cleanup = render(container, charm.asCell().key("$UI"));
+    // FIXME(ja): we don't have a cleanup function here!
```

Comment:
didn't this work?

```
const cleanup = render(container, charm.asCell().key("$UI"));
```

i'd expect React to call this every time a new charm is set, while `charm.asCell().key(UI)` takes care of reacting to any changes in the charm. and then you just get one cleanup function for the renderer.

Reply:
before changing to the effect code the event handlers wouldn't trigger (changing the title or adding a todo didn't trigger the handlers)

I found the effect code in llhl - perhaps there is a much smaller version - I can try!